cilium/CFP-41953-clustermesh-service-v2.md

```diff
@@ -0,0 +1,353 @@
+the loadbalancer k8s reflector. We expect clustermesh to achieve similar
+performance for service churn scenarios (100-3000x improvement over v1).
+Additionally, etcd operations are already rate limited (by default 20qps),
+which naturally coalesces multiple service updates in the workqueue,
+ensuring good latency and throughput without requiring explicit export
+throttling.
+
+Compressing the data with zstd will also dramatically reduce the on-wire size
+and etcd object size. We expect around 50x compression ratios or higher for
+services with thousands of backends. This significantly reduces control plane
+network bandwidth consumption.
+
+### Unifying ingestion pipelines through shared data structures
+
+The clustermesh and loadbalancer k8s reflector currently diverge in both
+their data structures and ingestion logic. This divergence has created the
+performance gap described in the Motivation section and makes it harder to
+maintain feature parity between the two code paths.
+
+We now propose unifying these pipelines by adopting shared Kubernetes data
+structures. Specifically, we will align clustermesh and the loadbalancer k8s
+reflector directly with Kubernetes slim `EndpointSlice` structs instead of
+using Cilium specific intermediate representations.
+
+The Kubernetes `EndpointSlice` API must remain backward compatible across
+Kubernetes versions. This aligns well with Cilium clustermesh's upgrade
+requirement to support at least two consecutive minor versions. Using the
+Kubernetes format directly also allows the loadbalancer code to use the
+same struct whether the data comes from a local Kubernetes cluster or from
+clustermesh.
+
+Cilium currently uses an internal `Endpoints` resource with the
+following struct:
+
+```go
+type Endpoints struct {
+        types.UnserializableObject
+        slim_metav1.ObjectMeta
+
+        EndpointSliceID
+
+        // Backends is a map containing all backend IPs and ports. The key to
+        // the map is the backend IP in string form. The value defines the list
+        // of ports for that backend IP, plus an additional optional node name.
+        // Backends map[cmtypes.AddrCluster]*Backend
+        Backends map[cmtypes.AddrCluster]*Backend
+}
+```
+
+This struct is a Cilium specific transformation of Kubernetes `EndpointSlice`
+data. It was designed to support both the legacy `Endpoints` API and the newer
+`EndpointSlice` API. With the introduction of statedb and most consumers watching
+statedb instead of relying on this resource, the only place actually using it
+is `operator/watchers/service_sync.go` to export service data for clustermesh.
+
+The proposal is thus to change this resource to directly expose the slim
+Kubernetes `EndpointSlice` struct and expose it in the same format in the
+new `ClusterServiceV2` struct. This avoids future divergence that could occur
+if the internal `Endpoints` struct is later optimized or changed for
+loadbalancer purposes.
```

Comment:
I am wondering if we shouldn't actually use a controller-runtime style controller to export EndpointSlice. Because we have a MCS-API controller that watches EndpointSlice, the future EndpointSlice sync controller to sync remote EndpointSlice and the service export side/service sync that would also watch that.

We could make all these things use controller-runtime style things in theory :thinking:. The resource could still be changed or even removed since there would be no consumer of it anymore afaik :thinking:.